service/openapi.yaml

```diff
@@ -0,0 +1,87 @@
+openapi: "3.0.0"
+info:
+  version: 0.1.0
+  title: OpenFeature Remote Evaluation Protocol (OFREP)
+  description: OFREP define the protocol for remote flag evaluations
+  contact:
+    url: https://github.com/open-feature/protocol
+  license:
+    name: Apache 2.0
+paths:
+  /evaluate:
+    post:
+      description: OFREP flag evaluation request
+      requestBody:
+        content:
+          'application/json':
+            schema:
+              $ref: '#/components/schemas/evaluationRequest'
+      responses:
+        '200':
+          description: OFREP successful evaluation response
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/success"
+        '400':
+          description: OFREP error response
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/failure"
+components:
+  schemas:
+    evaluationRequest:
+      description: Flag evaluation request
+      properties:
+        flagKey:
+          $ref: "#/components/schemas/flagKey"
+        flagType:
```

Comment:
Interesting discussion here. I think it really comes down to where do we put the responsibility of the type miss-matches getting resolved. Seeing that all the OpenFeature SDKs have typed flag methods `getBooleanValue()`, and that this API is basically a remote-evaluation interface for a Provider, and all the provider methods are typed `getStringEvaluation()` and expected to handle the type miss-matches, I think we should have these requests typed in some form. 

And seeing that we want to have a bulk evaluation endpoint, it doesn't make sense to have the API routes be individually typed and have to handle the bulk endpoint differently. Therefore the `evaluationRequest` should have a required type on it.

Reply:
> The concern I have is the knowledge of the type at flag management system. I can speak for flagd, where flag schema doesn't enforce flag type.

@Kavindu-Dodan 
But that's exactly why I do not see the need for a type in the request but rather in the response.
The flag management will have to have any knowledge of which type it has evaluated to. And maybe there is a case where it does not know for some reason but then the type in the response could be optional.
The provider can then do look at the type and try to parse the object value into the one specified in the response or declare a TYPE_MISMATCH if the evaluation method was called for another type.


> Seeing that all the OpenFeature SDKs have typed flag methods getBooleanValue(), and that this API is basically a remote-evaluation interface for a Provider, and all the provider methods are typed getStringEvaluation(), and the providers are expected to handle the type miss-matches, I think we should have these requests typed in some form.

I get the point.
But having the type in the request, or even type specific endpoints somewhat implies that flags could have more than one type. To me it would be way clearer to just evaluate a flag by a key and contextual information and then check the type based on the (optional) type in the response and the result of parsing as a consumer, which in our case is the provider implementation.


This is a long discussion and I am not sure how to get to a point here actually😅

Reply:
I tend to agree with @lukas-reining, especially since the bulk evaluation can be "evaluate everything", so no type of definition on the client.

And the flag management systems knows what are the types of the flags, so if the flag comes from the response it makes it explicit and the provider can just work with the type on his own.

Reply:
I'm in favor of not including the expected type in the request but it should be included in the response body. That would work with both bulk evaluations and individual evaluations.

Reply:
I am in favor of a type in the request. Primarily because I think the provider for a specific vendor should operate effectively the same with remote or local evaluation. Requests are made to the provider for specific types currently, so I think they also should be remotely.

In addition it also means that the provider has visibility. In the LaunchDarkly case each flag that is evaluated with a type that cannot be converted is also going to appear in analytics. It cannot appear if the provider isn't making that determination.

I am not sure what we are going to do about the evaluating all flags case. It seems to be a problem with both local and remote evaluation. I think the reasonable approach is to use an intermediate representation, most similar to a `structure`  for OF purposes or just an `any` type in languages that support it. All flags can be the type, and then the caller has to handle conversion (just keeping in mind that they are not always a 1-to-1 situation, especially numbers).

Reply:
In the community meeting on 29.02.2024, we decided not to include any type details in either request and response

This decision is now documented in a dedicated ADR [1]. We could add these details if we see the requirement in a future iteration. 

If no objections, I will resolve this conversation before merging the PR. 

[1] - https://github.com/open-feature/protocol/blob/0896456b001f217c53a155d6075854ef58ff06b2/service/adrs/0003-no-flag-type-in-request-or-response.md